File: Data Structures & Algorithms/invert-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        def dfs(cur_root):
            cur_root.left, cur_root.right = cur_root.right, cur_root.left
            if cur_root.left:
                dfs(cur_root.left)
            if cur_root.right:
                dfs(cur_root.right)
            
        dfs(root)
        
        return root

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    bool res = true;
    int bfs(TreeNode* curNode){
        if (curNode == nullptr){
            return 0;
        }
        int left = bfs(curNode->left);
        int right = bfs(curNode->right);
        if (abs(left - right) >= 2) {
            res = false;
        }
        return 1 + max(left, right);
    }
public:
    bool isBalanced(TreeNode* root) {
        bfs(root);
        return res;
    }
};
